rust/rope/src/compare.rs

```diff
@@ -0,0 +1,550 @@
+pub fn fast_cmpestr_mask(one: &[u8], two: &[u8]) -> i32 {
+    debug_assert_eq!(one.len(), 16);
+    debug_assert_eq!(two.len(), 16);
+    unsafe {
+        let onev = _mm_loadu_si128(one.as_ptr() as *const _);
+        let twov = _mm_loadu_si128(two.as_ptr() as *const _);
+        let mask = _mm_cmpestrm(onev, 16, twov, 16, SSID_OPTS);
+        _mm_movemask_epi8(mask)
+    }
+}
+
+/// Returns the lowest `i` for which `one[i] != two[i]`, if one exists.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn ne_idx(one: &[u8], two: &[u8]) -> Option<usize> {
+    let min_len = one.len().min(two.len());
+    let mut idx = 0;
+    loop {
+        let mask: i32;
+        // if slice is less than 16 bytes we manually pad it
+        if idx + SSE_STRIDE >= min_len {
+            let mut one_buf: [u8; 16] = [0; 16];
+            let mut two_buf: [u8; 16] = [0; 16];
+            let mut temp_idx = 0;
+            for i in idx..min_len {
+                one_buf[temp_idx] = one[i];
+                two_buf[temp_idx] = two[i];
+                temp_idx += 1;
+            }
+            mask = fast_cmpestr_mask(&one_buf, &two_buf);
+        } else {
+            mask = fast_cmpestr_mask(&one[idx..idx+SSE_STRIDE], &two[idx..idx+SSE_STRIDE]);
+        }
+        let i = mask.trailing_zeros() as usize;
+        if i != 32 { return Some(idx + i); }
+        idx += SSE_STRIDE;
+        if idx >= min_len { break; }
+    }
+    None
+}
+
+/// Returns the lowest `i` such that `one[one.len()-i] != two[two.len()-i]`,
+/// if one exists.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn ne_idx_rev(one: &[u8], two: &[u8]) -> Option<usize> {
+    let min_len = one.len().min(two.len());
+    let one = &one[one.len() - min_len..];
+    let two = &two[two.len() - min_len..];
+    debug_assert_eq!(one.len(), two.len());
+    let mut idx = min_len;
+    loop {
+        let mask: i32;
+        if idx < SSE_STRIDE {
+            let mut one_buf: [u8; 16] = [0; 16];
+            let mut two_buf: [u8; 16] = [0; 16];
+            let mut temp_idx = SSE_STRIDE - idx;
+            for i in 0..idx {
+                one_buf[temp_idx] = one[i];
+                two_buf[temp_idx] = two[i];
+                temp_idx += 1;
+            }
+            mask = fast_cmpestr_mask(&one_buf, &two_buf);
+        } else {
+            mask = fast_cmpestr_mask(&one[idx-SSE_STRIDE..idx], &two[idx-SSE_STRIDE..idx]);
+        }
+        let i = mask.leading_zeros() as usize - 16;
+        if i != 16 { return Some(min_len - (idx - i)); }
+        if idx < SSE_STRIDE { break; }
+        idx -= SSE_STRIDE;
+    }
+    None
+}
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+pub fn ne_idx(one: &[u8], two: &[u8]) -> Option<usize> {
+    ne_idx_fallback(one, two)
+}
+
+#[inline]
+#[allow(dead_code)]
+#[doc(hidden)]
+pub fn ne_idx_fallback(one: &[u8], two: &[u8]) -> Option<usize> {
+    for i in 0..one.len().min(two.len()) {
```

Comment:
This can be written something like `one.iter().zip(two.iter()).position(|a, b| a != b)`.

Reply:
interestingly I had that originally and it performed significantly worse on benchmarks, which I was surprised by. It's possible I was doing something wrong..